### PR TITLE
fnet cleanup part 1

### PR DESCRIPTION
**Pull request by ldemailly**
- report mode is now tested in Webtest
- use common code, now in fnet/ for listeners. add unit tests
- unified signature of various methods
- servers start are now all async
- fixes for localhost and port 0
- setup PPROF safely/explicitly and test in Webtest (that it's present in "server" and absent in "report" mode)

~based/stacked on top of #173 - might merge there but maybe it's easier to review separately~
now master based

### CODE REVIEW

File: fhttp/http.go [hunk 1]
```diff
@@ -395,21 +395,12 @@ func NewFastClient(o *HTTPOptions) Fetcher {
...
-	}
-	if len(addrs) > 1 && log.LogDebug() {
-		log.Debugf("Using only the first of the addresses for %s : %v", bc.host, addrs)
-	}
-	log.Debugf("Will go to %s", addrs[0])
-	bc.dest.IP = addrs[0]
-	bc.dest.Port, err = net.LookupPort("tcp", bc.port)
-	if err != nil {
-		log.Errf("Unable to resolve port '%s' : %v", bc.port, err)
+	addr := fnet.Resolve(bc.hostname, bc.port)
```
Comment (rkpagadala) [inline fhttp/http.go:398]: being pedantic :)
go idiom is to return err and check for it
```
- addr := fnet.Resolve(bc.hostname, bc.port)
- if addr == nil {
+ addr, err := fnet.Resolve(bc.hostname, bc.port)
+ if err != nil {
```

File: fhttp/http.go [hunk 2]
```diff
@@ -1198,35 +1189,42 @@ func closingServer(listener net.Listener) error {
...
+		err := s.Serve(listener)
+		if err != nil {
+			log.Fatalf("Unable to serve %s on %s: %v", name, addr.String(), err)
 		}
+	}()
+	return m, addr
+}
+
+// DynamicHTTPServer listens on an available port, sets up an http or https
+// (when secure is true) server on it and returns the listening port and
```
Comment (rkpagadala) [inline fhttp/http.go]: please update comment, no secure parameter

File: fhttp/http_test.go [hunk 3]
```diff
@@ -493,11 +493,11 @@ func TestGenerateSize(t *testing.T) {
 // Many of the earlier http tests are through httprunner but new tests should go here
 
 func TestEchoBack(t *testing.T) {
-	p, m := DynamicHTTPServer(false)
+	m, a := DynamicHTTPServer(false)
```
Comment (rkpagadala) [inline fhttp/http_test.go:496]:
```
-      m, a := DynamicHTTPServer(false)
+	mux, addr := DynamicHTTPServer(false)
```
here and elsewhere

File: ui/uihandler.go [hunk 4]
```diff
@@ -963,28 +976,22 @@ func RedirectToHTTPSHandler(w http.ResponseWriter, r *http.Request) {
...
-// sets the urlHostPort variable and returns hostport unmodified.
+// sets the urlHostPort variable.
 func setHostAndPort(inputPort string, addr *net.TCPAddr) {
 	urlHostPort = inputPort
 	portStr := inputPort
 	if addr != nil {
 		urlHostPort = addr.String()
 		portStr = fmt.Sprintf(":%d", addr.Port)
 	}
+	if !strings.Contains(inputPort, ":") {
```
Comment (rkpagadala) [inline ui/uihandler.go:992]: call NormalizePort?

File: fhttp/http.go (hunk 1 above)
Comment (ldemailly) [reply fhttp/http.go:398]: I know; but I think nil is also a perfectly valid marker for failure - with the assumption (which I realize is 'bad' for a library, I have a comment to that effect in fnet about splitting the functions / error handling - specially where I have log.Fatalf...) that error logging already happened

File: fhttp/http_test.go (hunk 3 above)
Comment (ldemailly) [reply fhttp/http_test.go:496]: I thought go was all about short names ? (those are test files)

File: fhttp/http.go (hunk 2 above)
Comment (ldemailly) [reply fhttp/http.go]: good catch, fixing

File: ui/uihandler.go (hunk 4 above)
Comment (ldemailly) [reply ui/uihandler.go:992]: yeah the logic is a bit in 2 places but I didn't want to re-call that method; I think I need to cleanup the whole uihandler some more in the future - right now just trying to get manual tests to work with localhost special case

File: fhttp/http_test.go (hunk 3 above)
Comment (rkpagadala) [reply fhttp/http_test.go:496]: not fixing is fine

File: ui/uihandler.go (hunk 4 above)
Comment (rkpagadala) [reply ui/uihandler.go:992]: ok

**Review by rkpagadala — APPROVED** [review]

**Pull request merged**